File: ruby/lib/data-structures/linked-list/doubly_linked_list.rb
require_relative 'node'
class DoublyLinkedList
  include Enumerable
  attr_reader :head, :tail
  def initialize head=nil
    @head = @tail = Node.new head unless head.nil?
  end

  def push_bottom node
    @tail = @head = Node.new(node) and return if @head.nil?
    @tail.next = Node.new(node)
    @tail.next.prev = @tail
    @tail = @tail.next
  end

  def push_top node
    node = Node.new(node)
    @tail = @head = node and return if @head.nil?
    node.next = @head
    @head.prev = node
    @head = node
  end

  def find data
    self.detect {|node| node.data == data }
  end

  def each
    return @head if @head.nil?
    current_node = @head
    until current_node.nil?
      yield current_node
      current_node = current_node.next
    end
  end

  def remove data
    @head = @head.next and return if @head.data == data
    node = self.find(data) || return
    next_node = node.next
    prev = node.prev
    next_node.prev = node.prev if next_node
    prev.next = node.next
  end

end

File: javascript/lib/data-structures/linked-list/doublyLinkedList.js
var Node = require('./node');

module.exports = DoublyLinkedList = function(){}

DoublyLinkedList.prototype = {
  addBottom: function(node) {
    if (this.head == undefined) return this.head = node;
    var currentNode = this.head;
    while(currentNode.next !== undefined) {
      currentNode = currentNode.next;
    }
    currentNode.next = node;
    node.prev = currentNode;
  },

  find: function(data) {
    var currentNode = this.head;
    while(currentNode !== undefined) {
      if(currentNode.data == data) return currentNode;
      currentNode = currentNode.next;
    }
  },

  addTop: function(node) {
    if (this.head == undefined) return this.head = node;
    node.next = this.head;
    this.head = node;
  },

  remove: function(data) {
    if (this.head.data == data) return this.head = this.head.next;
    var node = this.find(data);
    if (node === undefined) return;
    if(node.next !== undefined) node.next.prev = node.prev;
    node.prev.next = node.next;
    node.prev = node.next = undefined;

  }
}

